braindecode/augmentation/functional.py

```diff
@@ -969,3 +970,115 @@ def mixup(X, y, lam, idx_perm):
+    """Segment and reconstruct EEG data.
+
+    See [1]_ for details.
+
+    Parameters
+    ----------
+    X : torch.Tensor
+        EEG input example or batch.
+    y : torch.Tensor
+        EEG labels for the example or batch.
+    n_segments : int
+        Number of segments to use in the batch. If None, X will be
+        automatically segmented, getting the last element in a list
+        of factors of the number of samples's square root. Defaults to None.
+    random_state: int | numpy.random.Generator, optional
+        Used to draw the phase perturbation. Defaults to None.
+    Returns
+    -------
+    torch.Tensor
+        Transformed inputs.
+    torch.Tensor
+        Transformed labels.
+    References
+    ----------
+    .. [1] Lotte, F. (2015). Signal processing approaches to minimize or
+    suppress calibration time in oscillatory activity-based brain–computer
+    interfaces. Proceedings of the IEEE, 103(6), 871-890.
+    """
+
+    if y is not None:
+        if not isinstance(X, torch.Tensor) or not isinstance(y, torch.Tensor):
+            raise ValueError("X and y must be torch tensors.")
+
+        if X.shape[0] != y.shape[0]:
+            raise ValueError("Number of samples in X and y must be the same.")
+
+    if n_segments is None:
+        n_segments = int(X.shape[2])
+        n_segments_list = []
+        for i in range(1, int(n_segments**0.5) + 1):
+            if n_segments % i == 0:
+                n_segments_list.append(i)
+        n_segments = n_segments_list[-1]
+
+    elif not (isinstance(n_segments, (int, float)) and 1 <= n_segments <= X.shape[2]):
+        raise ValueError(
+            f"Number of segments must be a positive integer less than "
+            f"(or equal) the window size. Got {n_segments}"
+        )
+
+    # Initialize lists to store augmented data and corresponding labels
+    aug_data = []
+    aug_label = []
+
+    # Getting the random_state
+    rng = check_random_state(random_state)
+
+    if y is not None:
+        # Assuming 'y' is a tensor of labels, and 'X' is a tensor of data
+        classes = torch.unique(y)
+
+        data_classes = [(i, X[y == i]) for i in classes]
+        # Iterate through each class to separate and augment data
+        for class_index, X_class in data_classes:
+            # Determine class-specific dimensions
+            # Store the augmented data and the corresponding class labels
+            X_aug, n_trials = _segmentation_reconstruction(X_class, n_segments, rng)
+            aug_data.append(X_aug)
+            aug_label.append(
+                torch.full((n_trials,), class_index, dtype=y.dtype, device=y.device)
+            )
+    else:
+        X_aug, n_trials = _segmentation_reconstruction(X, n_segments, rng)
+        aug_data.append(X_aug)
+
+    # Concatenate the augmented data and labels
+    aug_data = torch.cat(aug_data, dim=0)
+
+    idx_shuffle = rng.permutation(len(aug_data))
+
+    aug_data = aug_data[idx_shuffle]
```

Comment:
Same here for the random opertation: could it be moved to the class level please?